src/gluonts/ev/interface.py

```diff
@@ -0,0 +1,191 @@
+from dataclasses import dataclass
+
+import numpy as np
+import pandas as pd
+
+from gluonts.dataset import DataEntry
+from gluonts.dataset.split import TestData
+from gluonts.ext.naive_2 import naive_2
+from gluonts.time_feature.seasonality import get_seasonality
+from gluonts.model.forecast import Forecast
+from gluonts.ev import seasonal_error
+
+
+@dataclass
+class ForecastDict:
+    """
+    Wrapper around ``Forecast`` objects, that adds ``__getitem__`` and
+    a batch dimension to returned arrays, for compatibility with ``gluonts.ev``.
+    """
+
+    forecast: Forecast
+
+    def __getitem__(self, name):
+        if name == "mean":
+            value = self.forecast.mean
+        elif name == "median":
+            value = self.forecast.median
+        else:
+            value = self.forecast.quantile(name)
+
+        return np.expand_dims(value, axis=0)
+
+
+def _get_data_batch(
+    input_: DataEntry, label: DataEntry, forecast: Forecast
+) -> ChainMap:
+    forecast_dict = ForecastDict(forecast)
+
+    freq = forecast.start_date.freqstr
+    seasonality = get_seasonality(freq=freq)
+
+    other_data = {
+        "label": np.array([label["target"]]),
+        "seasonal_error": np.array(
+            [seasonal_error(input_["target"], seasonality=seasonality)]
+        ),
+        "naive_2": np.array(
+            [naive_2(input_["target"], len(label["target"]), freq=freq)]
+        ),
+    }
+
+    return ChainMap(other_data, forecast_dict)
+
+
+def evaluate(
+    metrics, test_data: TestData, forecasts: Iterable[Forecast], axis=None
+) -> Tuple[dict, Optional[list]]:
+    """
+    Evaluate ``metrics`` by comparing ``forecasts`` to ``test_data``.
+
+    The optional ``axis`` arguments controls aggregation of the metrics:
+    - ``None`` (default) aggregates across all dimensions
+    - ``0`` aggregates across the batch dimension, giving one metric value per forecast horizon
+    - ``1`` aggregates across the time dimension, giving one metric value per batch item
+
+    Return results as a dictionary, plus an optional index (list).
+    """
+    evaluators = {}
+    for metric in metrics:
+        evaluator = metric(axis=axis)
+        evaluators[evaluator.name] = evaluator
+
+    assert axis is None or axis in [0, 1]
+    index = []
+    forecast_length = -1
+
+    for input_, label, forecast in zip(
+        test_data.input, test_data.label, forecasts
+    ):
+        forecast_length = len(forecast.index)
+        if axis == 1:
+            index.append((forecast.item_id, forecast.start_date))
+        data_batch = _get_data_batch(input_, label, forecast)
+        for evaluator in evaluators.values():
+            evaluator.update(data_batch)
+
+    if axis == 0:
+        index = list(range(forecast_length))
+    elif axis is None:
+        index = None
+
+    metrics_dict = {
+        metric_name: evaluator.get()
+        for metric_name, evaluator in evaluators.items()
+    }
+
+    return metrics_dict, index
```

Comment:
I don't like the `index` being returned from this function. Can we change it to only return the metric and return the index only when asked for it (by the `_pd` method)?

Reply:
The thing is, it needs to be constructed, and you may need to iterate through the data to do that (which is what `evaluate` does). So this is why it's returned there.

Another option is to provide it as part of `metrics_dict` itself, let me try that out.

Reply:
Part of the metrics dict would be better imo.

Reply:
Done, I agree it's better. One question is what the index name should be though, I guess it makes sense to have a different name if it's predicition horizon vs item_id + forecast time